hedera-node/src/main/java/com/hedera/services/state/virtual/schedule/ScheduleEqualityVirtualValue.java

```diff
@@ -0,0 +1,218 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.google.common.base.MoreObjects;
+import com.swirlds.common.io.streams.SerializableDataInputStream;
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import com.swirlds.virtualmap.VirtualValue;
+
+import java.beans.Transient;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+public class ScheduleEqualityVirtualValue implements VirtualValue {
+
+	static final int CURRENT_VERSION = 1;
+
+	static final long RUNTIME_CONSTRUCTABLE_ID = 0x1fe377366e3282f2L;
+
+	private final Map<String, Long> ids;
+
+	private boolean immutable;
+
+
+	public ScheduleEqualityVirtualValue() {
+		this(HashMap::new);
+	}
+
+	public ScheduleEqualityVirtualValue(Map<String, Long> ids) {
+		this(() -> new HashMap<>(ids));
+	}
+
+	private ScheduleEqualityVirtualValue(Supplier<Map<String, Long>> ids) {
+		this.ids = ids.get();
+	}
+
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || ScheduleEqualityVirtualValue.class != o.getClass()) {
+			return false;
+		}
+
+		var that = (ScheduleEqualityVirtualValue) o;
+		return Objects.equals(this.ids, that.ids);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hashCode(ids);
+	}
+
+	@Override
+	public String toString() {
+		var helper = MoreObjects.toStringHelper(ScheduleEqualityVirtualValue.class)
+				.add("ids", ids);
+		return helper.toString();
+	}
+
+	@Override
+	public void deserialize(SerializableDataInputStream in, int version) throws IOException {
+		int s = in.readInt();
+		ids.clear();
+		for (int x = 0; x <	s; ++x) {
+			byte[] keyBytes = new byte[in.readInt()];
+			in.readFully(keyBytes);
+			var key = new String(keyBytes, StandardCharsets.UTF_8);
+			ids.put(key, in.readLong());
+		}
+	}
+
+	@Override
+	public void deserialize(ByteBuffer in, int version) throws IOException {
+		int s = in.getInt();
+		ids.clear();
+		for (int x = 0; x <	s; ++x) {
+			byte[] keyBytes = new byte[in.getInt()];
+			in.get(keyBytes);
+			var key = new String(keyBytes, StandardCharsets.UTF_8);
+			ids.put(key, in.getLong());
+		}
+	}
+
+	@Override
+	public void serialize(SerializableDataOutputStream out) throws IOException {
+		out.writeInt(ids.size());
+		for (var e : ids.entrySet()) {
```

Comment:
⚠️ `HashMap` iteration is non-deterministic in the presence of garbage collection and/or VM restarts. We need to either sort the keys before serializing them; or use a `SortedMap` implementation.

Reply:
Good catch!

Reply:
fixed